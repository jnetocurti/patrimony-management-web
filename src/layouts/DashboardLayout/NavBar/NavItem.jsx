import React, { useState, Fragment } from 'react';
import { NavLink as RouterLink } from 'react-router-dom';

import {
  Button,
  Collapse,
  List,
  ListItem,
  makeStyles
} from '@material-ui/core';
import { ExpandLess, ExpandMore } from '@material-ui/icons';

import clsx from 'clsx';
import PropTypes from 'prop-types';

const useStyles = makeStyles((theme) => ({
  item: {
    display: 'flex',
    paddingTop: theme.spacing(0),
    paddingBottom: theme.spacing(0)
  },
  nested: {
    display: 'flex',
    paddingTop: theme.spacing(0),
    paddingBottom: theme.spacing(0),
    paddingLeft: theme.spacing(2)
  },
  button: {
    color: theme.palette.text.secondary,
    fontWeight: theme.typography.fontWeightMedium,
    justifyContent: 'flex-start',
    letterSpacing: theme.spacing(0),
    padding: '8px 6px',
    textTransform: 'none',
    width: '100%'
  },
  icon: {
    marginRight: theme.spacing(1)
  },
  title: {
    marginRight: 'auto'
  },
  active: {
    color: theme.palette.primary.main,
    '& $title': {
      fontWeight: theme.typography.fontWeightMedium
    },
    '& $icon': {
      color: theme.palette.primary.main
    }
  }
}));

const NavItem = ({ config }) => {
  const classes = useStyles();
  const [open, setOpen] = useState(config.open || false);

  const handleClick = () => setOpen(!open);

  const createItem = (options) => {
    const Icon = options.icon;
    const itemClass = options.isParent ? classes.item : classes.nested;
    return (
      <>
        <ListItem
          disableGutters
          className={clsx(itemClass)}
          onClick={options.handleClick}
        >
          <Button
            to={options.href}
            component={RouterLink}
            className={classes.button}
            activeClassName={classes.active}
          >
            {Icon && <Icon className={classes.icon} size="20" />}
            <span className={classes.title}>{options.title}</span>
          </Button>
          {options.isParent
            && options.subItems
            && (open ? <ExpandLess /> : <ExpandMore />)}
        </ListItem>
        {options.subItems && (
          <Collapse in={open} timeout="auto" unmountOnExit>
            <List component="div" disablePadding>
              {options.subItems.map((item) => (
                <Fragment key={item.title}>
                  {createItem({
                    isParent: false,
                    href: item.href,
                    icon: item.icon,
                    title: item.title
                  })}
                </Fragment>
              ))}
            </List>
          </Collapse>
        )}
      </>
    );
  };

  return (
    <>
      {createItem({
        handleClick,
        href: config.href,
        icon: config.icon,
        title: config.title,
        subItems: config.subItems,
        isParent: !!config.subItems
      })}
    </>
  );
};

NavItem.propTypes = {
  config: PropTypes.shape({
    open: PropTypes.bool,
    href: PropTypes.string.isRequired,
    title: PropTypes.string.isRequired,
    icon: PropTypes.elementType.isRequired,
    subItems: PropTypes.arrayOf(
      PropTypes.shape({
        href: PropTypes.string.isRequired,
        title: PropTypes.string.isRequired,
        icon: PropTypes.elementType.isRequired
      })
    )
  })
};

export default NavItem;
