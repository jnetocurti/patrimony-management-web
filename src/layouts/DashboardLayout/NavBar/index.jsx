import React, { useEffect } from 'react';
import { useLocation } from 'react-router-dom';

import {
  Box,
  Divider,
  Drawer,
  Hidden,
  List,
  makeStyles
} from '@material-ui/core';
import { Apartment as ApartmentIcon, ListOutlined } from '@material-ui/icons';

import PropTypes from 'prop-types';

import NavItem from './NavItem';

const items = [
  {
    open: true,
    icon: ApartmentIcon,
    title: 'Fundos imobiliários',
    href: '/app/real-estate-funds/ranking',
    subItems: [
      {
        icon: ListOutlined,
        title: 'Ranking',
        href: '/app/real-estate-funds/ranking'
      },
      {
        icon: ListOutlined,
        title: 'Transações',
        href: '/app/real-estate-funds/transactions'
      }
    ]
  }
];

const useStyles = makeStyles(() => ({
  mobileDrawer: {
    width: 225
  },
  desktopDrawer: {
    top: 64,
    width: 225,
    height: 'calc(100% - 64px)'
  }
}));

const NavBar = ({ onMobileClose, openMobile }) => {
  const classes = useStyles();
  const location = useLocation();

  useEffect(() => {
    if (openMobile && onMobileClose) {
      onMobileClose();
    }
  }, [location.pathname]);

  const content = (
    <Box height="100%" display="flex" flexDirection="column">
      <Divider />
      <Box p={2}>
        <List>
          {items.map((item) => (
            <NavItem key={item.title} config={item} />
          ))}
        </List>
      </Box>
      <Box flexGrow={1} />
    </Box>
  );

  return (
    <>
      <Hidden lgUp>
        <Drawer
          anchor="left"
          classes={{ paper: classes.mobileDrawer }}
          onClose={onMobileClose}
          open={openMobile}
          variant="temporary"
        >
          {content}
        </Drawer>
      </Hidden>
      <Hidden mdDown>
        <Drawer
          anchor="left"
          classes={{ paper: classes.desktopDrawer }}
          open
          variant="persistent"
        >
          {content}
        </Drawer>
      </Hidden>
    </>
  );
};

NavBar.propTypes = {
  onMobileClose: PropTypes.func,
  openMobile: PropTypes.bool
};

NavBar.defaultProps = {
  onMobileClose: () => {},
  openMobile: false
};

export default NavBar;
