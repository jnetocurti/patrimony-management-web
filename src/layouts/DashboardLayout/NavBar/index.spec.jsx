import React from 'react';
import { BrowserRouter } from 'react-router-dom';

import { mount } from 'enzyme';

import NavBar from './index';

describe('src/layouts/DashboardLayout/NavBar', () => {
  const onMobileClose = jest.fn();

  it('should render correctly openMobile false', () => {
    const openMobile = true;
    onMobileClose.mockReturnValue(false);

    const component = mount(
      <BrowserRouter>
        <NavBar {...{ onMobileClose, openMobile }} />
      </BrowserRouter>
    );
    expect(component).toMatchSnapshot();
  });

  it('should render correctly openMobile true', () => {
    const openMobile = false;
    onMobileClose.mockReturnValue(true);

    const component = mount(
      <BrowserRouter>
        <NavBar {...{ onMobileClose, openMobile }} />
      </BrowserRouter>
    );
    expect(component).toMatchSnapshot();
  });
});
