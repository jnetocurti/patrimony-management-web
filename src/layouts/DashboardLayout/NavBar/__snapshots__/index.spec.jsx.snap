// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`src/layouts/DashboardLayout/NavBar should render correctly openMobile false 1`] = `
<BrowserRouter>
  <Router
    action="POP"
    location={
      Object {
        "hash": "",
        "key": "default",
        "pathname": "/",
        "search": "",
        "state": null,
      }
    }
    navigator={
      Object {
        "action": "POP",
        "back": [Function],
        "block": [Function],
        "createHref": [Function],
        "forward": [Function],
        "go": [Function],
        "listen": [Function],
        "location": Object {
          "hash": "",
          "key": "default",
          "pathname": "/",
          "search": "",
          "state": null,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <NavBar
      onMobileClose={
        [MockFunction] {
          "calls": Array [
            Array [],
          ],
          "results": Array [
            Object {
              "type": "return",
              "value": false,
            },
          ],
        }
      }
      openMobile={true}
    >
      <Hidden
        lgUp={true}
      >
        <WithWidth(HiddenJs)
          lgDown={false}
          lgUp={true}
          mdDown={false}
          mdUp={false}
          smDown={false}
          smUp={false}
          xlDown={false}
          xlUp={false}
          xsDown={false}
          xsUp={false}
        />
      </Hidden>
      <Hidden
        mdDown={true}
      >
        <WithWidth(HiddenJs)
          lgDown={false}
          lgUp={false}
          mdDown={true}
          mdUp={false}
          smDown={false}
          smUp={false}
          xlDown={false}
          xlUp={false}
          xsDown={false}
          xsUp={false}
        />
      </Hidden>
    </NavBar>
  </Router>
</BrowserRouter>
`;

exports[`src/layouts/DashboardLayout/NavBar should render correctly openMobile true 1`] = `
<BrowserRouter>
  <Router
    action="POP"
    location={
      Object {
        "hash": "",
        "key": "default",
        "pathname": "/",
        "search": "",
        "state": null,
      }
    }
    navigator={
      Object {
        "action": "POP",
        "back": [Function],
        "block": [Function],
        "createHref": [Function],
        "forward": [Function],
        "go": [Function],
        "listen": [Function],
        "location": Object {
          "hash": "",
          "key": "default",
          "pathname": "/",
          "search": "",
          "state": null,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <NavBar
      onMobileClose={[MockFunction]}
      openMobile={false}
    >
      <Hidden
        lgUp={true}
      >
        <WithWidth(HiddenJs)
          lgDown={false}
          lgUp={true}
          mdDown={false}
          mdUp={false}
          smDown={false}
          smUp={false}
          xlDown={false}
          xlUp={false}
          xsDown={false}
          xsUp={false}
        />
      </Hidden>
      <Hidden
        mdDown={true}
      >
        <WithWidth(HiddenJs)
          lgDown={false}
          lgUp={false}
          mdDown={true}
          mdUp={false}
          smDown={false}
          smUp={false}
          xlDown={false}
          xlUp={false}
          xsDown={false}
          xsUp={false}
        />
      </Hidden>
    </NavBar>
  </Router>
</BrowserRouter>
`;
